pkg/controllers/teammembership/teammembership_updater_controller.go

```diff
@@ -0,0 +1,346 @@
+		Named(name).
+		For(&greenhousev1alpha1.TeamMembership{}).
+		Watches(&greenhousev1alpha1.Organization{},
+			handler.EnqueueRequestsFromMapFunc(r.enqueueTeamMembershipsForOrganization),
+			builder.WithPredicates(r.getPredicateFuncsForOrganization()),
+		).
+		Watches(&greenhousev1alpha1.Team{},
+			handler.EnqueueRequestsFromMapFunc(r.enqueueTeamMembershipsForTeam),
+			builder.WithPredicates(r.getPredicateFuncsForTeam()),
+		).
+		Complete(r)
+}
+
+func (r *TeamMembershipUpdaterController) enqueueTeamMembershipsForOrganization(ctx context.Context, obj client.Object) []reconcile.Request {
+	request := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name: obj.GetName(),
+			// Organization is cluster-scoped, it doesn't exist in namespaces.
+			Namespace: obj.GetName(),
+		},
+	}
+	return []reconcile.Request{request}
+}
+
+func (r *TeamMembershipUpdaterController) enqueueTeamMembershipsForTeam(ctx context.Context, obj client.Object) []reconcile.Request {
+	request := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      obj.GetName(),
+			Namespace: obj.GetNamespace(),
+		},
+	}
+	return []reconcile.Request{request}
+}
+
+func (r *TeamMembershipUpdaterController) getPredicateFuncsForOrganization() predicate.TypedFuncs[client.Object] {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			organization, ok := e.Object.(*greenhousev1alpha1.Organization)
+			if !ok {
+				return false
+			}
+			return organization.Spec.MappedOrgAdminIDPGroup != ""
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			organizationOld, ok := e.ObjectOld.(*greenhousev1alpha1.Organization)
+			if !ok {
+				return false
+			}
+			organizationNew, ok := e.ObjectNew.(*greenhousev1alpha1.Organization)
+			if !ok {
+				return false
+			}
+			return organizationOld.Spec.MappedOrgAdminIDPGroup !=
+				organizationNew.Spec.MappedOrgAdminIDPGroup
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return true
+		},
+	}
+}
+
+func (r *TeamMembershipUpdaterController) getPredicateFuncsForTeam() predicate.TypedFuncs[client.Object] {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			team, ok := e.Object.(*greenhousev1alpha1.Team)
+			if !ok {
+				return false
+			}
+			return team.Spec.MappedIDPGroup != ""
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			teamOld, ok := e.ObjectOld.(*greenhousev1alpha1.Team)
+			if !ok {
+				return false
+			}
+			teamNew, ok := e.ObjectNew.(*greenhousev1alpha1.Team)
+			if !ok {
+				return false
+			}
+			return teamOld.Spec.MappedIDPGroup !=
+				teamNew.Spec.MappedIDPGroup
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return true
+		},
+	}
+}
+
+// Lists all available teams und updates or creates respective teamMemberships if team.Spec.MappedIDPGroup is present
+func (r *TeamMembershipUpdaterController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	teamList := greenhousev1alpha1.TeamList{}
```

Comment:
The `ctrl.Request` object coming in here, will always refer to `name` and `namespace` of a `Team` and therefore of a `TeamMembership`. It suffices to reconcile this one team.